content/en/blog/2022/instrument-kafka-clients/index.md

```diff
@@ -0,0 +1,252 @@
+
+```java
+Resource resource = Resource.getDefault()
+        .merge(Resource.create(Attributes.of(ResourceAttributes.SERVICE_NAME, "my-kafka-service")));
+
+SdkTracerProvider sdkTracerProvider = SdkTracerProvider.builder()
+        .addSpanProcessor(BatchSpanProcessor.builder(JaegerGrpcSpanExporter.builder().build()).build())
+        .setSampler(Sampler.alwaysOn())
+        .setResource(resource)
+        .build();
+
+OpenTelemetry openTelemetry = OpenTelemetrySdk.builder()
+        .setTracerProvider(sdkTracerProvider)
+        .setPropagators(ContextPropagators.create(W3CTraceContextPropagator.getInstance()))
+        .buildAndRegisterGlobal();
+```
+
+### Using the interceptors
+
+The Kafka clients API provides a way to "intercept" messages before they are sent to the brokers as well as messages received from the broker before being passed to the application.
+This approach is heavily used when you need to add some logic or content to the messages right before they are sent.
+At same time, it is useful to handle consumed messages right before they are passed to the upper application layer.
+It fits pretty well with tracing when you need to create or close spans on sending and receiving messages.
+
+The Kafka clients instrumentation library provides two interceptors to be configured to add tracing information automatically.
+The interceptor classes have to be set in the properties bag used to create the Kafka client within the application.
+
+Use the `TracingProducerInterceptor` for the producer in order to create a "send" span automatically, each time a message is sent.
+
+```java
+props.setProperty(ProducerConfig.INTERCEPTOR_CLASSES_CONFIG, TracingProducerInterceptor.class.getName());
+```
+
+Use the `TracingConsumerInterceptor` for the consumer in order to create a "receive" span automatically, each time a message is received.
+
+```java
+props.setProperty(ConsumerConfig.INTERCEPTOR_CLASSES_CONFIG, TracingConsumerInterceptor.class.getName());
+```
+
+### Wrapping the clients
+
+The other way is by wrapping the Kafka client with a tracing enabled one.
+
+On the producer side, assuming you have a `Producer<K, V>` instance, you can wrap it in the following way.
+
+```java
+KafkaTelemetry telemetry = KafkaTelemetry.create(GlobalOpenTelemetry.get());
+Producer<String, String> tracingProducer = telemetry.wrap(producer);
+```
+
+Then use the `tracingProducer` as usual for sending messages to the Kafka cluster.
+
+On the consumer side, assuming you have a `Consumer<K, V>` instance, you can wrap it in the following way.
+
+```java
+KafkaTelemetry telemetry = KafkaTelemetry.create(GlobalOpenTelemetry.get());
+Consumer<String, String> tracingConsumer = telemetry.wrap(this.consumer);
+```
+
+Then use the `tracingConsumer` as usual for receiving messages from the Kafka cluster.
+
+## Instrumenting by using the agent
```

Comment:
I personally would like seeing that automatic way first, people (including myself) are lazy, so having something they can do within a few "clicks" is always a better hook than copying a bunch of code snippets.

Reply:
That's a good point. I will re-arrange the order!

Reply:
Done!